hostfactory/host_provider/src/allocation_strategy.py

```diff
@@ -0,0 +1,224 @@
+from math import ceil, floor
+from enum import Enum
+import logging
+
+
+class AllocationStrategies(Enum):
+    PRICE = "price"
+    CAPACITY = "capacity"
+    DECAY = "decay"
+    WEIGHTED = "weighted"
+
+
+def round_down_to_nearest_multiple(target, x):
+    return x * floor(target / x)
+
+def round_up_to_nearest_multiple(target, x):
+    return x * ceil(target / x)
+
+def calculate_vm_dist_capacity(vm_types, total_slot_count):
+    vm_dist = {}
+    slot_counts = [total_slot_count // len(vm_types) for n in range(len(vm_types))]
+
+    remaining_slots = total_slot_count
+    vm_dist = {vm: 0 for vm in vm_types.keys()}
+    for n, (vm, weight) in enumerate(vm_types.items()):        
+        vm_dist[vm] = round_down_to_nearest_multiple(slot_counts[n], weight)
+        remaining_slots -= vm_dist[vm]
+        if remaining_slots <= 0:
+            break
+
+    if remaining_slots > 0:
+        for n, (vm, weight) in enumerate(vm_types.items()):
+            vm_dist[vm] += weight
+            remaining_slots -= weight
+            if remaining_slots <= 0:
+                break
+
+    logging.info("New (capacity) VM SKU distribution targets: %s", vm_dist)
+    return vm_dist
+
+
+def calculate_vm_dist_weighted(vm_types, requested_slot_count, percentage_weights=[.7, .2, .05, .05]):
+    '''Apply a simple static percentage distribution to the vm_types'''
+    vm_dist = {vm: 0 for vm in vm_types.keys()}
```

Comment:
We should assert that the weights actually come out to 1, no?
abs(sum(percentage_weights) - 1.0) < 1e10
or something

Reply:
good point.